### PR TITLE
PMM-12910: Popup new version available

### PR DESCRIPTION
**Pull request by doracretu3pillar**
https://perconadev.atlassian.net/browse/PMM-12910

### CODE REVIEW

File: public/app/percona/shared/services/updates/Updates.service.ts [hunk 1]
```diff
@@ -1,8 +1,21 @@
...
 import { CheckUpdatesBody, CheckUpdatesResponse } from './Updates.types';
 
 export const UpdatesService = {
   getCurrentVersion: (body: CheckUpdatesBody = { force: false }) =>
-    api.post<CheckUpdatesResponse, CheckUpdatesBody>('/v1/Updates/Check', body, true),
+    api.get<CheckUpdatesResponse, CheckUpdatesBody>('/v1/server/updates', false, { params: body }),
+
+  getUpdatesChangelogs: () => api.get<CheckUpdatesChangeLogsResponse, void>('/v1/server/updates/changelogs', false),
+
+  setSnoozeCurrentVersion: (body: SnoozePayloadBody) =>
```
Comment (matejkubinec) [inline public/app/percona/shared/services/updates/Updates.service.ts:17]: APIs for `snooze` are already part of the `User.service`, no need to add them here.

I think the PUT `/v1/users/me` can be updated partially (just send field for the snooze version)

File: public/app/percona/shared/core/reducers/updates/updates.types.ts [hunk 2]
```diff
@@ -26,3 +28,29 @@ export interface CheckUpdatesPayload {
...
+
+export interface SnoozePayloadBody {
+  productTourCompleted: boolean;
+  alertingTourCompleted: boolean;
+  snoozedPmmVersion: string;
+}
+
+export interface SnoozePayloadResponse {
+  userId: number;
+  productTourCompleted: boolean;
```
Comment (matejkubinec) [inline public/app/percona/shared/core/reducers/updates/updates.types.ts:53]: Doesn't the backend return data in snake_case?

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.styles.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  `,
+  version: css`
+    margin-bottom: ${spacing.md};
+  `,
+  releaseNotesText: css`
+    a {
+      color: ${colors.textBlue};
+    }
+  `,
+  howToUpdateTitle: css``,
```
[the remark this answers is not in the document]
Comment (doracretu3pillar) [reply public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.styles.ts:20]: fixed

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.test.tsx [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+import * as GrafanaUpdates from 'app/percona/shared/core/reducers/updates/updates';
+import { UpdatesService } from 'app/percona/shared/services/updates';
+import { configureStore } from 'app/store/configureStore';
+
+import PerconaUpdateVersion from './PerconaUpdateVersion';
+
+const checkUpdatesChangeLogsSpy = jest.spyOn(GrafanaUpdates, 'checkUpdatesChangeLogs');
+describe('PerconaUpdateVersion', () => {
+  function setup(store) {
```
[the remark this answers is not in the document]
Comment (doracretu3pillar) [reply public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.test.tsx]: fixed

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.test.tsx [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                releaseNotesText: 'text1',
+              },
+            ],
+          },
+        },
+      },
+    };
+    jest
+      .spyOn(UpdatesService, 'getUpdatesChangelogs')
+      .mockReturnValue(Promise.resolve({ ...state.percona.updates.changeLogs }));
```
[the remark this answers is not in the document]
Comment (doracretu3pillar) [reply public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.test.tsx]: the values from the mocked object will be set on the payload then assigned in checkUpdatesChangeLogs.fulfilled

image

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        title={Messages.titleMultipleUpdates}
+        isOpen={showUpdate && changeLogs && changeLogs?.updates?.length > 1}
+        className={styles.updateVersionModal}
+      >
+        <div data-testid="multiple-updates-modal">
+          <h5 className={styles.newVersionsTitle}>{Messages.newVersions}</h5>
+          <ul className={styles.listOfReleaseNotes}>
+            {changeLogs?.updates.map((update: UpdatesChangelogs) => (
+              <li key={update.toString()}>
+                <a href={update.releaseNotesUrl}>
```
[the remark this answers is not in the document]
Comment (doracretu3pillar) [reply public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx:107]: fixed on both modals

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.test.tsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import { UpdatesService } from 'app/percona/shared/services/updates';
+import { configureStore } from 'app/store/configureStore';
+
+import PerconaUpdateVersion from './PerconaUpdateVersion';
+import { StoreState } from 'app/types';
+import { EnhancedStore } from '@reduxjs/toolkit';
+
+const checkUpdatesChangeLogsSpy = jest.spyOn(GrafanaUpdates, 'checkUpdatesChangeLogs');
+describe('PerconaUpdateVersion', () => {
+  function setup(store: EnhancedStore) {
```
[the remark this answers is not in the document]
Comment (doracretu3pillar) [reply public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.test.tsx:16]: it doesn't work in all the cases, for example:

    _const store = configureStore({
      ...defaultState,
      percona: {
        ...defaultState.percona,
        ...state,
      },
    } as StoreState);_
    
    here I need to have my _state_ inside _percona_
    
it's going to be overly complicated to have a unique setup in your case

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        const differenceInMilliseconds = currentDate.getTime() - lastCheckDate.getTime();
+        return Math.trunc(differenceInMilliseconds / (1000 * 60 * 60 * 24));
+      } else {
+        return 7;
+      }
+    };
+
+    const prepareModal = async () => {
+      if (!snoozeCurrentVersion) {
+        await dispatch(getSnoozeCurrentVersion());
```
[the remark this answers is not in the document]
Comment (doracretu3pillar) [reply public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx:31]: It's ok. Only if updateAvailable = true, I want getSnoozeCurrentVersion to be called. Then, the useEffect will be called again, and show/not show the modal based on snoozed.

Calling a useless API (useless time consumption) VS entering once again in useEffect.

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+      <Modal
+        onDismiss={onDismiss}
+        title={Messages.titleOneUpdate}
+        isOpen={showUpdate && changeLogs && changeLogs?.updates?.length === 1}
+        className={styles.updateVersionModal}
+      >
+        <div data-testid="one-update-modal">
+          <h5 className={styles.version}>{changeLogs?.updates[0]?.version || ''}</h5>
+          <p className={styles.releaseNotesText}>
+            <a href={changeLogs?.updates[0]?.releaseNotesUrl || ''}>{Messages.fullReleaseNotes}</a>
```
[the remark this answers is not in the document]
Comment (doracretu3pillar) [reply public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx:82]: fixed

File: public/app/percona/shared/services/updates/Updates.service.ts (hunk 1 above)
Comment (doracretu3pillar) [reply public/app/percona/shared/services/updates/Updates.service.ts:17]: fixed

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.test.tsx [hunk 10]
```diff
@@ -13,7 +12,6 @@ import { StoreState } from 'app/types';
 import { EnhancedStore } from '@reduxjs/toolkit';
 
 const checkUpdatesChangeLogsSpy = jest.spyOn(GrafanaUpdates, 'checkUpdatesChangeLogs');
-const fetchUserDetailsActionSpy = jest.spyOn(UserTourReducer, 'fetchUserDetailsAction');
 
 describe('PerconaUpdateVersion', () => {
```
Comment (matejkubinec) [inline public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.test.tsx:16]: I think a test case for the snooze functionality (that the button calls the api) would be helpful.

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx [hunk 11]
```diff
@@ -62,7 +57,7 @@ const PerconaUpdateVersion: FC = () => {
         className={styles.updateVersionModal}
       >
         <div data-testid="one-update-modal">
-          <h5 className={styles.version}>{changeLogs?.updates[0]?.version || ''}</h5>
+          <h5 className={styles.version}>{latest?.version || ''}</h5>
           <p className={styles.releaseNotesText}>
             <a target="_blank" rel="noopener noreferrer" href={changeLogs?.updates[0]?.releaseNotesUrl || ''}>
```
Comment (matejkubinec) [inline public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx:62]: I think you can `latest` here also. (if it contains the releaseNotesUrl, not 100% sure)

Comment (doracretu3pillar) [reply public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx:62]: fixed

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaBootstrapper.tsx [hunk 12]
```diff
@@ -104,49 +109,54 @@ export const PerconaBootstrapper = ({ onReady }: PerconaBootstrapperProps) => {
...
-          <HorizontalGroup justify="flex-end" spacing="md">
-            <Button variant="secondary" onClick={finishTour}>
-              {Messages.skip}
-            </Button>
-            <Button variant="secondary" onClick={() => setModalIsOpen(false)}>
-              {Messages.checkLater}
-            </Button>
-          </HorizontalGroup>
-        </Modal>
+      {updateAvailable && showUpdate ? (
```
Comment (matejkubinec) [inline public/app/percona/shared/components/PerconaBootstrapper/PerconaBootstrapper.tsx:112]: I think there's a missing chaeck for the update loading to be false.

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+import { checkUpdatesChangeLogs, UpdatesChangeLogs } from 'app/percona/shared/core/reducers/updates';
+import { setSnoozedVersion } from 'app/percona/shared/core/reducers/user/user';
+import { getPerconaUser, getUpdatesInfo } from 'app/percona/shared/core/selectors';
+import { useAppDispatch } from 'app/store/store';
+import { useSelector } from 'app/types';
+
+import { Messages } from './PerconaUpdateVersion.constants';
+import { getStyles } from './PerconaUpdateVersion.styles';
+
+const PerconaUpdateVersion = ({ showUpdate, setShowUpdate }: PerconaUpdateVersionProps) => {
```
Comment (matejkubinec) [inline public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx:16]: I think you can move the state to the updates store, then you could access the show modal state anywhere in the tree.

Comment (doracretu3pillar) [reply public/app/percona/shared/components/PerconaBootstrapper/PerconaUpdateVersion/PerconaUpdateVersion.tsx]: fixed

File: public/app/percona/shared/components/PerconaBootstrapper/PerconaBootstrapper.tsx (hunk 12 above)
Comment (doracretu3pillar) [reply public/app/percona/shared/components/PerconaBootstrapper/PerconaBootstrapper.tsx]: fixed

**Review by matejkubinec — APPROVED** [review]

**Review by matejkubinec — APPROVED** [review]

**Review by matejkubinec — APPROVED** [review]

**Pull request merged**